playbooks/gcp/openshift-cluster/install.yml

```diff
@@ -0,0 +1,27 @@
+# This playbook installs onto a provisioned cluster
+---
+- hosts: localhost
+  connection: local
+  tasks:
+  - name: place all scale groups into Ansible groups
+    include_role:
+      name: openshift_gcp
+      tasks_from: setup_scale_group_facts.yml
+
+- name: configure the control plane
```

Comment:
just make this ../deploy_cluster.yml and add install_gcp.yml to deploy_cluster.yml in the appropriate spot.  You can use a boolean set via inventory or via this playbook to ensure that install_gcp.yml is called only when we want.

Reply:
We're probably going to want to add prerequisites.yml to this playbook.  I have been working on some patches that integrate prerequisites.yml into these types of special plays.

See this node-scaleup PR for an example: https://github.com/openshift/openshift-ansible/pull/6784

Reply:
The question is - why does prerequisites have to be special, vs being part of lower level playbooks?  Is it only the dependency on openshift_facts?

Reply:
It's special because it incorporates some things (like installing docker and configuring docker storage) there were previously not included in openshift-ansible and had to be completed by the user before running openshift-ansible.

container_runtime (previously the docker role) needs to be configured very early in the deployment process.  The reasons for this are:
1. configuring registry authentication credentials
2. openshift_version depends on correctly setup container runtime to query latest/available tags for containerized installs
3. health checks want to ensure containers are available and registries are configured before running the installer and failing halfway through.

We could put this role in the init code path, but certain users don't want us to touch docker after it's been installed, because as you may know, we have given users license to modify things outside of openshift-ansible, and re-running the container_runtime role may have undesired effects for them.  Additionally, it runs against the 'nodes' group, so having to run those tasks each time a play is run adds a lot of execution time to large clusters.

We could probably add it to deploy_cluster.yml now that we have refactored a lot of other components.  But I think we're a little late in the cycle for that for 3.9 and we could potentially do it in 3.10.

If you look at 3.7 (or even worse, 3.6), you will see that it was a very tangled web of dependencies surrounding the docker role.  I agree that prerequisites.yml is a compromise, might not be the best solution, but it works and is an improvement over past releases.

Reply:
> This is bad. This shouldn't be a dependency then.

If you compare what we have now versus 3.7 and 3.6 releases you'll see a vast reduction in the web of dependencies in this area.  I agree, we can definitely do more in this area, but I think we're in a pretty good place at the moment compared to previous releases in this area.  Personally, I would like to force users to specify exactly what they want and we take the training wheels off in regards to openshift_version, but that's not something we implemented this cycle.  I think we have the ground work now to move that direction if we choose to.

> We either need to avoid the dependency on the container runtime to do the check, or figure out how to make those checks special and move them somewhere else.

container_runtime needs to be setup at some point regardless.  We might as well do it right up front and check to make sure things are working before we start deploying to 100's of nodes.  Whether we're using atomic, skopeo, or docker to interact with registries, we still need to ensure those items are properly configured on the hosts.  I don't see a reason to not configure container_runtime up front when it's a prerequisite to pretty much the whole deployment.

Reply:
> We might as well do it right up front and check to make sure things are working before we start deploying to 100's of node

This is the old way :)  In the cloud provider model, we will need to check these things *before* we have a single node provisioned, because we are going to be using golden images.  Anything that we need in order to check this needs to be something that can be installed into the docker image that the playbook runs out of, and it's better for it not to be tied in any way to a particular container runtime choice (whether skopeo or docker) because the component won't know that.

I think the checks are valuable, but we can't rely on installing a container runtime on a given provisioned node in order to the checks.  We probably shouldn't be installing skopeo either, given the amount of dependencies it pulls in, or its relatively inflexibility.  The checks won't make any sense in cloud providers really, because we won't even invoke ansible on the nodes during the install (i.e. both AWS and GCP really operate only on the nodes selected as masters).

Reply:
At best, this seems like a check that has to be done after container runtime is installed in the node job, not up front in the playbooks.

Reply:
> This is the old way :) In the cloud provider model, we will need to check these things before we have a single node provisioned, because we are going to be using golden images. 

Building the golden images is currently no different on aws vs deploying on bare metal, as far as container_runtime is concerned.  If container runtime is not setup, then the node isn't provisioned; if there's no container runtime, it can't run containers.

We can't really check 'before' a single node is provisioned; the value is in checking that each node is configured correctly (ie, networking, proxy), not so much that you put in valid credentials and registry url.

We have a lot of checks built in for user friendliness.  Probably you or I or anyone that is commit PRs here can identify most problems somewhat quickly; The average user just sees the nodes didn't deploy and don't know why.  Having the checks in place that we do gives them much better feedback if they are not advanced ansible users (I would argue, many of our users are not 'advanced' ansible users, they just want the installer to work).